[dtensor] slice_backward to use op strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130289
* #130288
* __->__ #130287
* #130286

as titled. slice_backward right now forward the sharding
unconditionally, which is wrong mathmatically. This PR switch it to op
strategy and only allow replication

cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[dtensor] slice_backward to use op strategy

as titled. slice_backward right now forward the sharding
unconditionally, which is wrong mathmatically. This PR switch it to op
strategy and only allow replication

[ghstack-poisoned]

awgu commented:
> slice_backward right now forward the sharding
unconditionally, which is wrong mathmatically

Would it be worth to include a unit test covering a case that was wrong before this PR?

wanchaol committed:
Update on "[dtensor] slice_backward to use op strategy"


as titled. slice_backward right now forward the sharding
unconditionally, which is wrong mathmatically. This PR switch it to op
strategy and only allow replication

[ghstack-poisoned]

wanchaol commented:
> Would it be worth to include a unit test covering a case that was wrong before this PR?

added one unit test and verified that it was failed before

awgu commented:
@pytorchbot merge

awgu commented:
@pytorchbot merge
